### DISCUSSION TITLE
How to setup a relay node

### DISCUSSION THREAD

**Discussion by deeptuuk** [libp2p/rust-libp2p#2252] [General]
hello

i want to setup a relay node , but i don't know how to do it,
Does anybody give me some suggestion，thank you

there is three node A，B，C
A and C under the NAT, B has public IP, A want dial C directly.

I browser the rust doc libp2p about relay section, but i get stuck with it!

Please help me, thank you !

here is my code
```
use libp2p::core::transport::memory::MemoryTransport;
use libp2p::relay::{RelayConfig, new_transport_and_behaviour};
use libp2p::swarm::{Swarm, SwarmEvent};
use libp2p::core::{identity, Multiaddr, multiaddr::Protocol, PeerId, upgrade, Transport};
use libp2p::yamux::YamuxConfig;
use libp2p::plaintext::PlainText2Config;
use libp2p::tcp::TcpConfig;
use std::convert::TryInto;
use std::str::FromStr;

use futures::prelude::*;
use futures::executor::block_on;
use futures::StreamExt;
use std::error::Error;
use std::net::Ipv4Addr;
use std::task::Poll;

fn main() -> Result<(), Box<dyn Error>> {

    let local_key = identity::Keypair::generate_ed25519();
    let local_peer_id = PeerId::from(local_key.public());
    println!("Local peer id: {:?}", local_peer_id);

    let plaintext = PlainText2Config {
        local_public_key: local_key.public().clone(),
    };    

    let (relay_transport, relay_behaviour) = new_transport_and_behaviour(
        RelayConfig::default(),
        TcpConfig::new(),
    );

    
    let transport = relay_transport
        .upgrade(upgrade::Version::V1)
        .authenticate(plaintext)
        .multiplex(YamuxConfig::default())
        .boxed();
    
    let mut swarm = Swarm::new(transport, relay_behaviour, local_peer_id);

    let relay_addr = Multiaddr::empty()
        .with(Protocol::Ip4(Ipv4Addr::new(0, 0, 0, 0))) // Relay address.
        .with(Protocol::Tcp(0))
        .with(Protocol::P2p(*local_peer_id.as_ref())) // Relay peer id.
        .with(Protocol::P2pCircuit); // Signal to listen via remote relay node.    

    
    swarm.listen_on(relay_addr)?;

    
    block_on(future::poll_fn(move |cx| loop {
        match swarm.poll_next_unpin(cx) {
            Poll::Ready(Some(event)) => println!("{:?}", event),
            Poll::Ready(None) => return Poll::Ready(()),
            Poll::Pending => return Poll::Pending
        }
    }));    

    Ok(())
}
```

**Comment by mxinden** [comment]
Hi @deeptuuk,

**Circuit relay v1**

For the v1 relay protocol you can follow https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/examples/relay.rs for both dialing client, listening client and relay server.

**Circuit relay v2**

The v2 implementation is not yet merged (https://github.com/libp2p/rust-libp2p/pull/2059/). Still you can use https://github.com/mxinden/rust-libp2p-server/ as a server and https://github.com/libp2p/rust-libp2p/blob/b26d89f6982725cf60a0d8adfe3284c218725839/protocols/dcutr/examples/client.rs as a dialing client and as a listening client.

Let us know if the above helps ;)

**Reply by deeptuuk** [reply]
Thank you so much， Let me try it！

**Reply by deeptuuk** [reply]
hi @mxinden 
I want know wether AutoNat is in progress，thank you

**Reply by mxinden** [reply]
@elenaf9 is currently looking into implementing AutoNAT. I suggest you follow https://github.com/libp2p/rust-libp2p/issues/2052 where we will post any status updates.

**Reply by r-zig** [reply]
Hi @deeptuuk as @mxinden wrote in the https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/examples/relay.rs example exiting a docker compose that make it simple to run all 3 nodes at once with different networks
See 
https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/examples/docker-compose.yml
